### PR TITLE
support mig devices

### LINKED ISSUE #3: Nomad plugin nvidia-gpu does not detect multi-instance GPUs
### Nomad version
`Nomad v1.2.0-dev (6d35e2fb58663aa2ad8b8f47459eff342901e72a)`

_(The nightly build that fixed hashicorp/nomad#11342)_

### Operating system and Environment details

```
# cat /etc/redhat-release
Red Hat Enterprise Linux release 8.4 (Ootpa)
# cat /proc/cpuinfo | grep EPYC | uniq
model name      : AMD EPYC 7763 64-Core Processor
# cat /proc/meminfo | grep -i Memtot
MemTotal:       527815668 kB
# nvidia-smi -L
GPU 0: NVIDIA A100-SXM4-40GB (UUID: GPU-X-Y-Z-T-F)
GPU 1: NVIDIA A100-SXM4-40GB (UUID: GPU-X-Y-Z-T-F)
GPU 2: NVIDIA A100-SXM4-40GB (UUID: GPU-X-Y-Z-T-F)
GPU 3: NVIDIA A100-SXM4-40GB (UUID: GPU-X-Y-Z-T-F)
```
### Issue

Hello Again !

First of all, thank you very much for the amazing response time and time to fix the preemption problem in hashicorp/nomad#11342.
While I was testing your product, I wondered about its compatibility with the [Multi-Instance GPU solution of nVidia](https://docs.nvidia.com/datacenter/tesla/mig-user-guide/).

In a nutshell, it allows for us to physically partition a big GPU into more bite-sized GPUs. That can be immensly useful for numerous use cases _(e.g. hosting multiple jobs on the same chonkster of a GPU)_.

When MIG is enabled on a GPU, you cannot use the GPU as a resource (i.e. You can only use the MIG instances created on the GPU). For example, in my setup we have 4 A100 GPUs, of which GPU3 has MIG enabled. I've went ahead and created 2 half-GPUs _(basically)_. This is the `nvidia-smi -L`:

```
GPU 0: NVIDIA A100-SXM4-40GB (UUID: GPU-X-Y-Z-T-P)
GPU 1: NVIDIA A100-SXM4-40GB (UUID: GPU-X-Y-Z-T-P)
GPU 2: NVIDIA A100-SXM4-40GB (UUID: GPU-X-Y-Z-T-P)
GPU 3: NVIDIA A100-SXM4-40GB (UUID: GPU-X-Y-Z-T-P)
  MIG 3g.20gb     Device  0: (UUID: MIG-X-Y-Z-T-P)
  MIG 3g.20gb     Device  1: (UUID: MIG-X-Y-Z-T-P)
```

Yet when I run `nomad node status {my_node_id}`, this is the output of the GPU resources part:

```
Device Resource Utilization
nvidia/gpu/NVIDIA A100-SXM4-40GB[GPU-X-Y-Z-T-P]  3 / 40536 MiB
nvidia/gpu/NVIDIA A100-SXM4-40GB[GPU-X-Y-Z-T-P]  0 / 40536 MiB
nvidia/gpu/NVIDIA A100-SXM4-40GB[GPU-X-Y-Z-T-P]  39340 / 40536 MiB
nvidia/gpu/NVIDIA A100-SXM4-40GB[GPU-X-Y-Z-T-P]  20 / 40536 MiB
```

Now this is problematic for two reasons:
1. Should nomad schedule a workload on GPU3, then that workload will fail as the resource isn't effectively available.
2. Nomad cannot recognize MIG instances, although they are basically distinct resources, and that in turn restricts us to for example allocating a 40GB GPU to a job that otherwise would only need 10GB VRAM and 10% of the CUDA cores of the beefy GPU.

Now MIG instances are usable _(almost as fully-fledged GPUs, or more specifically CUDA devices)_, so they're perfectly compatible with Docker workloads using the nVidia runtime _(e.g. instead of using --gpus='"device=0"' you'd use --gpus='"device=0:0"' to reference the first MIG instance on the first GPU)_.

### Reproduction steps

1. Enable MIG by following [the nVidia guide](https://docs.nvidia.com/datacenter/tesla/mig-user-guide/).
2. Relaunch the Nomad agent.

#### Expected Result

1. See the GPU with MIG enabled disappear from the available resources (as its no longer usable)
2. See the new created MIG instances appear as resources that can be assigned to Docker jobs.

#### Actual Result

1. The GPU w/ MIG enabled is still there in the available resources.
2. No MIG instances are picked up.

### PR DESCRIPTION
**Pull request by shoenig**
Incorporates previous PRs by @attachmentgenie and @isidentical while fixing a couple of bugs and adding MIG specific tests to the mock driver implementation.

- **Add rudimentary support for MIGs**
- **Record utilizations from parent device**
- **feat: switch to _v2 calls for memory infra and reintroducing ecc info. Adding some enhanced error catching for nvml queries**
- **chore: bumping to go 1.22**
- **chore: introduce MIG support**
- **driver: fixup more MIG feature bugs**

Closes #3
Closes #27 
Closes #40

### CODE REVIEW

**Comment by shoenig** [comment]
spot check:

```
$ nvidia-smi -L
GPU 0: NVIDIA A100-SXM4-40GB (UUID: GPU-d8d7e984-7a52-6ad0-676c-222f4be482b9)
GPU 1: NVIDIA A100-SXM4-40GB (UUID: GPU-658c3ec8-b842-8f3d-96b8-e919d699d17f)
GPU 2: NVIDIA A100-SXM4-40GB (UUID: GPU-143664cc-e9e6-dea8-c036-a112f4f4cf2e)
GPU 3: NVIDIA A100-SXM4-40GB (UUID: GPU-41c37062-3644-9e82-f1e5-4cf9832c5294)
GPU 4: NVIDIA A100-SXM4-40GB (UUID: GPU-b2210207-2f27-05df-a93d-8d060db4eabb)
GPU 5: NVIDIA A100-SXM4-40GB (UUID: GPU-6109dd5c-d378-24ee-f344-79b537574ea6)
  MIG 1g.5gb      Device  0: (UUID: MIG-8c4c5a27-e508-5556-b2ea-7e549465741e)
  MIG 1g.5gb      Device  1: (UUID: MIG-1d89abd8-d10a-5361-bd06-77a807048472)
  MIG 1g.5gb      Device  2: (UUID: MIG-c6b63f1b-a006-5803-b4f4-cc1b2867b34b)
  MIG 1g.5gb      Device  3: (UUID: MIG-0f03f623-f28d-5831-bc51-f450dad99b8c)
  MIG 1g.5gb      Device  4: (UUID: MIG-5915594e-2939-5009-89d2-cc81c198bd6f)
  MIG 1g.5gb      Device  5: (UUID: MIG-108eed9d-d296-5037-b840-0a54cccdc3fe)
  MIG 1g.5gb      Device  6: (UUID: MIG-705077ed-ed26-5909-aacd-4842d47b8c76)
GPU 6: NVIDIA A100-SXM4-40GB (UUID: GPU-c2d35f3f-98bc-7b5e-bf23-5b82ebd9a30e)
GPU 7: NVIDIA A100-SXM4-40GB (UUID: GPU-c7488922-4511-39b7-102b-552925241e95)
  MIG 3g.20gb     Device  0: (UUID: MIG-3dbd382d-dc25-5c9f-b925-e2b3c1b1513b)
  MIG 2g.10gb     Device  1: (UUID: MIG-d7128c1b-5510-54ea-9b5f-96975eb31882)
  MIG 1g.5gb      Device  2: (UUID: MIG-d12b4c16-9fd6-522c-8232-3e45bb4cdbfe)
  MIG 1g.5gb      Device  3: (UUID: MIG-7bca62a9-3c21-5ee3-99ab-008cd3f47af5)
```

```
$ nomad node status 52 | grep '^nvidia/gpu'
nvidia/gpu/NVIDIA A100-SXM4-40GB MIG 1g.5gb[MIG-0f03f623-f28d-5831-bc51-f450dad99b8c]   <none>
nvidia/gpu/NVIDIA A100-SXM4-40GB MIG 1g.5gb[MIG-108eed9d-d296-5037-b840-0a54cccdc3fe]   <none>
nvidia/gpu/NVIDIA A100-SXM4-40GB MIG 1g.5gb[MIG-1d89abd8-d10a-5361-bd06-77a807048472]   <none>
nvidia/gpu/NVIDIA A100-SXM4-40GB MIG 1g.5gb[MIG-5915594e-2939-5009-89d2-cc81c198bd6f]   <none>
nvidia/gpu/NVIDIA A100-SXM4-40GB MIG 1g.5gb[MIG-705077ed-ed26-5909-aacd-4842d47b8c76]   <none>
nvidia/gpu/NVIDIA A100-SXM4-40GB MIG 1g.5gb[MIG-7bca62a9-3c21-5ee3-99ab-008cd3f47af5]   <none>
nvidia/gpu/NVIDIA A100-SXM4-40GB MIG 1g.5gb[MIG-8c4c5a27-e508-5556-b2ea-7e549465741e]   <none>
nvidia/gpu/NVIDIA A100-SXM4-40GB MIG 1g.5gb[MIG-c6b63f1b-a006-5803-b4f4-cc1b2867b34b]   <none>
nvidia/gpu/NVIDIA A100-SXM4-40GB MIG 1g.5gb[MIG-d12b4c16-9fd6-522c-8232-3e45bb4cdbfe]   <none>
nvidia/gpu/NVIDIA A100-SXM4-40GB MIG 2g.10gb[MIG-d7128c1b-5510-54ea-9b5f-96975eb31882]  <none>
nvidia/gpu/NVIDIA A100-SXM4-40GB MIG 3g.20gb[MIG-3dbd382d-dc25-5c9f-b925-e2b3c1b1513b]  <none>
nvidia/gpu/NVIDIA A100-SXM4-40GB[GPU-143664cc-e9e6-dea8-c036-a112f4f4cf2e]              633 / 40960 MiB
nvidia/gpu/NVIDIA A100-SXM4-40GB[GPU-41c37062-3644-9e82-f1e5-4cf9832c5294]              633 / 40960 MiB
nvidia/gpu/NVIDIA A100-SXM4-40GB[GPU-658c3ec8-b842-8f3d-96b8-e919d699d17f]              633 / 40960 MiB
nvidia/gpu/NVIDIA A100-SXM4-40GB[GPU-d8d7e984-7a52-6ad0-676c-222f4be482b9]              633 / 40960 MiB
```

File: nvml/driver_linux.go [hunk 1]
```diff
@@ -159,70 +233,106 @@ func buildID(id [32]int8) string {
...
-	utzMem := uint(utz.Memory)
-
-	utzEnc, _, code := nvml.DeviceGetEncoderUtilization(device)
+	mem, code := nvml.DeviceGetMemoryInfo(device)
 	if code != nvml.SUCCESS {
-		return nil, nil, decode("failed to get device encoder utilization", code)
+		return nil, nil, decode("failed to get device memory utilization", code)
 	}
-	utzEncU := uint(utzEnc)
+	memUsedU := mem.Used / (1 << 20)
```
Comment (schmichael) [inline nvml/driver_linux.go:252]: Comment and/or a const for `1 << 20`

**Review by schmichael — APPROVED** [review]
:tada:

File: nvml/client.go [hunk 2]
```diff
@@ -156,23 +167,32 @@ func (c *nvmlClient) GetStatsData() ([]*StatsData, error) {
...
+	for uuid, mode := range deviceUUIDs {
+
+		// A30/A100 MIG devices have no stats.
+		//
+		// https://docs.nvidia.com/datacenter/tesla/mig-user-guide/#telemetry
+		//
+		// Is this fixed on H100 or later? Maybe?
+		if mode == mig || mode == parent {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (shoenig) [reply nvml/client.go:188]: If we log we end up spamming a log line for each MIG device for each period. In the sad case that's 7 MIG devices for 8 GPUs every 30 seconds which is a lot of log spam for the hope Nvidia will fix their stuff.

**Pull request merged**